### ISSUE TITLE
Concepts 'more constrained' not detected for member functions

### ISSUE THREAD

**Issue by royjacobson** [llvm/llvm-project#56154]
https://godbolt.org/z/bvKGhY4Yz

```c++
template <int N> concept C0 = (N == 0);
template <int N, int N2> concept C0x = C0<N>;
template <int N1, int N2> concept C00 = C0x<N1, N2> && C0<N2>;

template<int N1, int N2>
struct A {
    void f() requires C00<N1, N2>;
    void f() requires C0x<N1, N2>;
};

template<int N1, int N2>
void f() requires C00<N1, N2>;

template<int N1, int N2>
void f() requires C0x<N1, N2>;

int main() {
    f<0, 0>(); // Works

    A<0, 0> a;
    a.f();  // Works in GCC only.
}
```

This should compile, I think, but it doesn't. For some reason MSVC also say that `a.f()` is ambiguous.

**Comment by yuanfang-chen** [comment]
Just did a search and found the very interesting PR (https://reviews.llvm.org/D119544) you mentioned. Yeah, I think that should fix this bug.

**Comment by royjacobson** [comment]
So, after some debugging, this looks like a bug in `NormalizedConstraint::fromConstraintExpr`. When normalizing the constraint it replaces the template argument with `DeclRefExpr ... 'int' NonTypeTemplateParm ... 'N' 'int'` that doesn't preserve the arguments substitution chain and then we think `C0x` subsumes `C00`.

But the normalization code is pretty dense so I haven't made progress on how to fix it.

**Comment by usx95** [comment]
On the similar lines:
Redefinition of function is accepted but this looks like an ODR violation.
```
template <int N> concept C0 = (N == 0);
template <int N, int N2> concept C0x = C0<N>;
template <int N1, int N2> concept C00 = C0x<N1, N2> && C0<N2>;

template<int N1, int N2>
int f() requires C00<N1, N2> { return 0; }

template<int N1, int N2>
int f() requires C0x<N1, N2> { return 1; }

int main() {
    return f<0, 0>(); 
}
```
https://godbolt.org/z/KMrnWzx56
